libc/time/utimes.c

```diff
@@ -0,0 +1,470 @@
+ *    - lutimes()
+ *
+ * Copyright 2022-2024 Phoenix Systems
+ * Author: Ziemowit Leszczynski, Mateusz Bloch
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <unistd.h>
+#include <fcntl.h>
+#include <time.h>
+#include <errno.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/time.h>
+
+#include "unity_fixture.h"
+
+
+static const char *const FILENAME = "utimes";
+static const char *const LINKNAME = "utimes_link";
+static const unsigned int LOOP_CNT = 10;
+
+
+typedef enum {
+	TEST_SET_NULL,
+	TEST_SET_NOW,
+	TEST_SET_PAST,
+	TEST_SET_ONE,
+	TEST_SET_HIGH_VALUE
+} test_t;
+
+
+typedef enum {
+	UTIMES,
+	FUTIMES,
+	LUTIMES
+} mode_test;
+
+
+/* Function to set time values based on the test case */
+static void set_time_values(struct timeval tv[2], test_t test)
+{
+	switch (test) {
+		case TEST_SET_NULL:
+		case TEST_SET_NOW:
+			TEST_ASSERT_EQUAL_INT(0, gettimeofday(&tv[0], NULL));
+			tv[1] = tv[0];
+			break;
+		case TEST_SET_PAST:
+			tv[0].tv_sec = random();
+			tv[1].tv_sec = random();
+			tv[0].tv_usec = random() % 1000000;
+			tv[1].tv_usec = random() % 1000000;
+			break;
+		case TEST_SET_ONE:
+			tv[0].tv_sec = 1;
+			tv[1].tv_sec = 1;
+			tv[0].tv_usec = 1;
+			tv[1].tv_usec = 1;
+			break;
+		case TEST_SET_HIGH_VALUE:
+			tv[0].tv_sec = INT32_MAX;
+			tv[1].tv_sec = INT32_MAX;
+			tv[0].tv_usec = 999999;
+			tv[1].tv_usec = 999999;
+			break;
+		default:
+			TEST_FAIL_MESSAGE("Invalid test case");
+			break;
+	}
+}
+
+/* Function to perform utimes/futimes/lutimes based on the mode */
+static void perform_utimes(const char *filename, struct timeval tv[2], test_t test, mode_test mode, struct stat *statbuf, char *message)
+{
+	int fd;
+
+	if (mode == UTIMES) {
+		strcpy(message, "utimes() failed");
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, utimes(filename, test == TEST_SET_NULL ? NULL : tv), message);
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, stat(filename, statbuf), message);
+	}
+	else if (mode == FUTIMES) {
+		strcpy(message, "futimes() failed");
+		TEST_ASSERT_GREATER_THAN_INT(0, fd = open(filename, O_RDONLY));
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, futimes(fd, test == TEST_SET_NULL ? NULL : tv), message);
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, fstat(fd, statbuf), message);
+		close(fd);
+	}
+	else if (mode == LUTIMES) {
+		strcpy(message, "lutimes() failed");
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, lutimes(filename, test == TEST_SET_NULL ? NULL : tv), message);
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, lstat(filename, statbuf), message);
+	}
+	else {
+		TEST_FAIL_MESSAGE("Invalid mode");
+	}
+}
+
+/* Function to validate the time values after performing utimes/futimes/lutimes */
+static void validate_time_values(const struct timeval tv[2], const struct stat *statbuf, test_t test, char *message)
+{
+	int64_t adiff_sec, mdiff_sec;
+#ifndef __phoenix__
+	int64_t adiff_nsec, mdiff_nsec;
+#endif
+
+	adiff_sec = tv[0].tv_sec - statbuf->st_atim.tv_sec;
```

Comment:
It should be `statbuf->st_atim.tv_sec - tv[0].tv_sec`, please note that  stat is called later than getting tv.
But to not accept every value of statbuf lower than tv I'd suggest abs, so:
```suggestion
	adiff_sec = abs(statbuf->st_atim.tv_sec - tv[0].tv_sec);
```
and change accordingly the rest of diff values